Bump iOS deployment target from 13.0 to 14.0

mokagio opened the pull request:
Bump iOS deployment target from 13.0 to 14.0.

## To test

The app should build locally. Ideally, there should be no deprecation warnings or errors.

~~Currently, we have a bunch of errors due to "Today View extensions have been deprecated". The reason we have an error and not a warning is because those targets have "Treat Warnings as Errors" set to `YES`.~~ Update: #19509 addressed those. 

<details>
<summary>See Today View errors</summary>

![Untitled](https://user-images.githubusercontent.com/1218433/197158238-aec9626b-5b69-4703-8e28-3cc97ddd3f18.jpg)

</details>

## Regression Notes

1. Potential unintended areas of impact – N.A.
2. What I did to test those areas of impact (or what existing automated tests I relied on) – N.A.
3. What automated tests I added (or what prevented me from doing so) – N.A.

---

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes. **N.A.**
- [x] I have considered adding accessibility improvements for my changes. **N.A.**
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary. **N.A.**

mokagio committed:
Remove target-level `IPHONEOS_DEPLOYMENT_TARGET` setting

This way, all the targets will use the same iOS deployment, all coming
from the project-level setting.

mokagio committed:
Extract `IPHONEOS_DEPLOYMENT_TARGET` to `SDKVersions.xcconfig`

mokagio committed:
Bump `IPHONEOS_DEPLOYMENT_TARGET` from 13.0 to 14.0

mokagio committed:
Add `RELEASE-NOTES.txt` entry for dropping iOS 13

staskus commented:
All iOS13 widget targets `WordPressAllTimeWidget`, `WordPressThisWeekWidget`, `WordPressTodayWidget` can be removed since we have iOS14+ `WordPressStatsWidgets` as a replacement.

guarani commented:
If there are no objections to dropping iOS 13 support and it's possible to merge this PR by Friday, we could:
1. Target v21.1 with this PR
2. Remove ~~Home~~ **_Today_** View widgets in this PR to remove the errors
3. Rebase https://github.com/wordpress-mobile/WordPress-iOS/pull/19479 (which also targets 21.1) onto this branch
4. Close https://github.com/wordpress-mobile/WordPress-iOS/pull/19490

From your side @mokagio, does this seem like a viable approach?
If it's not possible to merge this by Friday this week (e.g. if there are more errors to fix), we may have to support iOS 13 widgets on Jetpack until this PR is merged.
Wdyt @staskus? Are there better alternatives?

mokagio commented:
@guarani yes sounds good. I started working on this (`mokagio/remove-ios-13-widgets`) and was aiming to get it done today, but I need to drop off unexpectedly. Stay tuned.

staskus commented:
> Rebase https://github.com/wordpress-mobile/WordPress-iOS/pull/19479 (which also targets 21.1) onto this branch

I think maybe we should merge iOS14+ to trunk and then adjust _this_ PR branch accordingly? Widgets changes are a must for 21.1, unlike a version bump.

> If it's not possible to merge this by Friday this week (e.g. if there are more errors to fix), we may have to support iOS 13 widgets on Jetpack until this PR is merged.
Wdyt @staskus? Are there better alternatives?

@guarani I don't think it makes sense to support iOS13 widgets, even if we won't manage to bump the deployment target this release. The usability of iOS13 is too low and we would be removing the changes next release either way.

mokagio committed:
Use `CLLocationManager` instance property for `authorizationStatus`

The class property was deprecated in iOS 14. Xcode gave the following
warning (which we treat as an error):

>  Deprecated. Use the instance property `authorizationStatus` instead.

mokagio committed:
Replace deprecated `CLLocationManagerDelegate` method

The deprecation warning in Xcode wasn't informative, but the deprecation
annotation suggested what to do:

```
/*
 *  locationManager:didChangeAuthorizationStatus:
 *
 *  Discussion:
 *    Invoked when the authorization status changes for this application.
 */
- (void)locationManager:(CLLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status API_DEPRECATED_WITH_REPLACEMENT("-locationManagerDidChangeAuthorization:", ios(4.2, 14.0), macos(10.7, 11.0), watchos(1.0, 7.0), tvos(9.0, 14.0));
```

mokagio committed:
Remove iOS 13 only `SchedulingCalendarViewController`

mokagio commented:
@guarani @staskus 

>> Rebase https://github.com/wordpress-mobile/WordPress-iOS/pull/19479 (which also targets 21.1) onto this branch

> I think maybe we should merge iOS14+ to trunk and then adjust this PR branch accordingly? Widgets changes are a must for 21.1, unlike a version bump.

The [iOS 13 widget removals](https://github.com/wordpress-mobile/WordPress-iOS/pull/19509) is ready for review, so there should be ample time for the Jetpack widgets to go on top of this.

> Widgets changes are a must for 21.1

Are they? Is there an external commitment that must be met by that version? _Just curious_.

guarani commented:
> I think maybe we should merge iOS14+ to trunk and then adjust _this_ PR branch accordingly? Widgets changes are a must for 21.1, unlike a version bump.

This is a good point, I think we can afford to ship iOS 14 Home Screen widgets on Jetpack, without trying to support iOS 13 Today View widgets on Jetpack as well. The only users who will benefit from iOS Today View widgets are:
- At most 0.05% of current JP app users (this is the percentage of users on iOS 13, but the number is lower because not all will install widgets).
- At most 0.5% of current WP users (this is the percentage of users on iOS 13 and assumes all switch to Jetpack)

> Are they? Is there an external commitment that must be met by that version? _Just curious_.

Not an external commitment, but an internal goal for the Jetpack app to offer the same features the WordPress app offers by Nov 17. Shipping widgets on Jetpack in 21.1 contributes to this goal.

mokagio commented:
@guarani @staskus prioritizing the Jetpack widgets sounds good to me. There are only two days left for that in my time zone by now and we'd be cutting it close any other way.

I'll wait for that to land on `trunk` then update everything accordingly. Let me know if there's something I can do on the Jetpack widgets end.

staskus commented:
@mokagio [Jetpack Widgets](https://github.com/wordpress-mobile/WordPress-iOS/pull/19479) merged ✅

mokagio commented on WordPress/Classes/ViewRelated/Post/Scheduling/SchedulingCalendarViewController.swift:
```diff
@@ -1,253 +0,0 @@
-import Gridicons
-
-// MARK: - Date Picker
-@available(iOS, deprecated: 14.0, message: "Use SchedulingDatePickerViewController, based on UIDatePicker.inline")
-class SchedulingCalendarViewController: UIViewController, CalendarSheet, DateCoordinatorHandler, SchedulingViewControllerProtocol {
```
The deprecation annotation recommends using `SchedulingDatePickerViewController` instead of `SchedulingCalendarViewController` in iOS 14 and above. 

Now that we support only iOS 14 and above, this whole file can be deleted. Notice that in this file we also had `CalendarSheet` and `TimePickerViewController`.

mokagio commented on WordPress/Classes/ViewRelated/Post/Scheduling/SchedulingViewControllerPresenter.swift:
```diff
@@ -64,6 +59,11 @@ class PresentableSchedulingViewControllerProvider: PresentableSchedulingViewCont
     }
 }
 
+// FIXME: This protocol is redundant as of dropping iOS 13.
+//
+// It was used as a facade in between `SchedulingCalendarViewController` (iOS 13) and
+// `SchedulingDatePickerViewController` (iOS 14+). `SchedulingCalendarViewController` has been
+// deleted so we can remove this as well.
 protocol SchedulingViewControllerProtocol: UIViewController, UIViewControllerTransitioningDelegate, UIAdaptivePresentationControllerDelegate {
```
I'll follow this up later on. I wanted to keep the scope limited.

mokagio commented on WordPress/WordPressTest/ViewRelated/Post/Controllers/TimePickerViewControllerTests.swift:
```diff
@@ -1,28 +0,0 @@
-
-class TimePickerViewControllerTests: XCTestCase {
-
-    var timePickerVC: TimePickerViewController?
```
`TimePickerViewController` was removed together with `

mokagio commented:
@crazytonyli I asked for your review on this PR with failing CI because I already have #19509 which addresses the issue and has been approved. I didn't want to merge it in, otherwise the diff would have become bigger.

Let me know what you think.

crazytonyli commented:
@mokagio Am I correct in saying that this PR is not longer needed? Can we change the target branch of #19509 from `drop-ios-13` to `trunk` and merge it?

mokagio commented:
> Can we change the target branch of #19509 from `drop-ios-13` to `trunk` and merge it?

Oh yes, that would do it too. I guess what I should have done was create the widgets PR from `trunk`. So that, once they were merged, rebasing/merging `trunk` into this one would have showed a green CI.

But, given that PR was based off this one, your suggestion will save us a merge. Thanks!

mokagio commented:
Closing in favor of merging #19509 directly into `trunk`
